python-test-samples/apigw-sqs-lambda-sqs/README.md

```diff
@@ -0,0 +1,155 @@
+* **Confirm changes before deploy**: If set to yes, SAM CLI shows you any change sets for manual review before deployment. If set to no, the AWS SAM CLI will automatically deploy application changes.
+* **Allow SAM CLI IAM role creation**: Many AWS SAM templates, including this example, create AWS IAM roles required for the AWS Lambda function(s) included to access AWS services. By default, SAM CLI scopes these down to minimum required permissions. To deploy an AWS CloudFormation stack which creates or changes IAM roles, the `CAPABILITY_IAM` value for `capabilities` must be provided. If you don't provide permission through this prompt, you must explicitly pass `--capabilities CAPABILITY_IAM` to the `sam deploy` command.
+* **CheckOutputQueue may not have authorization defined, Is this okay?**:set to yes. since this is for testing only we can proceed without client autoriztion for API GW, but for production enviroment we do recommand to have it.
+
+* **Save arguments to samconfig.toml**: If set to yes, SAM CLI saves your choices to a configuration file inside the project, so that in the future you can just re-run `sam deploy` without parameters to deploy changes to your application.
+
+You can find your API Gateway Endpoint URL in the output values displayed after deployment. Take note of this URL for use in the logging section below. On subsequent deploys you can run `sam deploy` without the `--guided` flag. [Read the documentation](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-deploying.html).
+
+[Top](#contents)
+
+---
+
+## Run the Integration Test
+[test_api_gateway.py](tests/integration/test_api_gateway.py) 
+
+For integration tests, it is assumed the full stack is already deployed before testing. if not, you can use the following commands to deploy (see [Prerequisites](#prerequisites) section):
+```shell
+apigw-sqs-lambda-sqs$ sam build
+apigw-sqs-lambda-sqs$ sam deploy --guided
+```
+
+The [integration test](tests/integration/test_api_gateway.py) setup determines the [API endpoint](tests/integration/test_api_gateway.py#L56).  
+
+The [integration test tear-down](tests/integration/test_api_gateway.py#L108) removes any data injected for the tests and purges each queue in case they have messages inside them.
+
+To run the integration test, create the environment variable "AWS_SAM_STACK_NAME" with the name of the test stack, and execute the test.
+
+```shell
+# Set the environment variables AWS_SAM_STACK_NAME and (optionally)AWS_DEFAULT_REGION 
+# to match the name of the stack and the region where you will test
+# pip install will check the requirements file and will install needed packages to run the pytest. it is required only once for setting the enviroment
+
+apigw-sqs-lambda-sqs$ export AWS_SAM_STACK_NAME=<stack-name>
+apigw-sqs-lambda-sqs$ pip install -r tests/requirements.txt 
+apigw-sqs-lambda-sqs$ python -m pytest -s tests/integration -v 
+```
```

Comment:
Test fail:

```
rootdir: /Users/wtromano/Documents/repo/github/serverless-test-samples/python-test-samples/apigw-sqs-lambda-sqs, configfile: pytest.ini
plugins: socket-0.5.1
collected 3 items                                                                                                                                                                                                           

tests/integration/test_api_gateway.py::TestApiGateway::test_api_gateway_200 FAILED
tests/integration/test_api_gateway.py::TestApiGateway::test_api_gateway_404 FAILED
tests/integration/test_api_gateway.py::TestApiGateway::test_dead_letter_queue FAILED

tests/integration/test_api_gateway.py::TestApiGateway::test_dead_letter_queue ERROR

========================================================================================================== ERRORS ===========================================================================================================
________________________________________________________________________________ ERROR at teardown of TestApiGateway.test_dead_letter_queue _________________________________________________________________________________

self = <class 'test_api_gateway.TestApiGateway'>

    def teardown_class(self) -> None:
        """
        # For tear-down, remove any data injected for the tests
        # purge each queue in case they have messages inside them
        # since purge can be done once every 60 sec, we start with 60 sec sleep
        """
        logging.info("Teardown Phase...")
        time.sleep(60)
    
        # cleaning input queue
>       client = boto3.client("sqs")

tests/integration/test_api_gateway.py:118: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/__init__.py:92: in client
    return _get_default_session().client(*args, **kwargs)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/session.py:299: in client
    return self._session.create_client(
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:951: in create_client
    credentials = self.get_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:509: in get_credentials
    ).load_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:2039: in load_credentials
    creds = provider.load()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:977: in load
    creds_dict = self._retrieve_credentials_using(credential_process)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:1004: in _retrieve_credentials_using
    parsed = botocore.compat.json.loads(stdout.decode('utf-8'))
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/__init__.py:346: in loads
    return _default_decoder.decode(s)
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:337: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x10c29b280>, s = '', idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:355: JSONDecodeError
========================================================================================================= FAILURES ==========================================================================================================
____________________________________________________________________________________________ TestApiGateway.test_api_gateway_200 ____________________________________________________________________________________________

self = <class 'test_api_gateway.TestApiGateway'>

    @classmethod
    def setUp(self) -> None:
        """
        Based on the provided env variable AWS_SAM_STACK_NAME,
        We use the cloudformation API to retrieve the APIGW URL
        """
        stack_name = TestApiGateway.get_stack_name()
    
>       cf_client = boto3.client("cloudformation")

tests/integration/test_api_gateway.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/__init__.py:92: in client
    return _get_default_session().client(*args, **kwargs)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/session.py:299: in client
    return self._session.create_client(
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:951: in create_client
    credentials = self.get_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:509: in get_credentials
    ).load_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:2039: in load_credentials
    creds = provider.load()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:977: in load
    creds_dict = self._retrieve_credentials_using(credential_process)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:1004: in _retrieve_credentials_using
    parsed = botocore.compat.json.loads(stdout.decode('utf-8'))
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/__init__.py:346: in loads
    return _default_decoder.decode(s)
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:337: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x10c29b280>, s = '', idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:355: JSONDecodeError
____________________________________________________________________________________________ TestApiGateway.test_api_gateway_404 ____________________________________________________________________________________________

self = <class 'test_api_gateway.TestApiGateway'>

    @classmethod
    def setUp(self) -> None:
        """
        Based on the provided env variable AWS_SAM_STACK_NAME,
        We use the cloudformation API to retrieve the APIGW URL
        """
        stack_name = TestApiGateway.get_stack_name()
    
>       cf_client = boto3.client("cloudformation")

tests/integration/test_api_gateway.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/__init__.py:92: in client
    return _get_default_session().client(*args, **kwargs)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/session.py:299: in client
    return self._session.create_client(
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:951: in create_client
    credentials = self.get_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:509: in get_credentials
    ).load_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:2039: in load_credentials
    creds = provider.load()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:977: in load
    creds_dict = self._retrieve_credentials_using(credential_process)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:1004: in _retrieve_credentials_using
    parsed = botocore.compat.json.loads(stdout.decode('utf-8'))
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/__init__.py:346: in loads
    return _default_decoder.decode(s)
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:337: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x10c29b280>, s = '', idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:355: JSONDecodeError
___________________________________________________________________________________________ TestApiGateway.test_dead_letter_queue ___________________________________________________________________________________________

self = <class 'test_api_gateway.TestApiGateway'>

    @classmethod
    def setUp(self) -> None:
        """
        Based on the provided env variable AWS_SAM_STACK_NAME,
        We use the cloudformation API to retrieve the APIGW URL
        """
        stack_name = TestApiGateway.get_stack_name()
    
>       cf_client = boto3.client("cloudformation")

tests/integration/test_api_gateway.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/__init__.py:92: in client
    return _get_default_session().client(*args, **kwargs)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/session.py:299: in client
    return self._session.create_client(
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:951: in create_client
    credentials = self.get_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:509: in get_credentials
    ).load_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:2039: in load_credentials
    creds = provider.load()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:977: in load
    creds_dict = self._retrieve_credentials_using(credential_process)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:1004: in _retrieve_credentials_using
    parsed = botocore.compat.json.loads(stdout.decode('utf-8'))
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/__init__.py:346: in loads
    return _default_decoder.decode(s)
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:337: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x10c29b280>, s = '', idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:355: JSONDecodeError
================================================================================================== short test summary info ==================================================================================================
FAILED tests/integration/test_api_gateway.py::TestApiGateway::test_api_gateway_200 - json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
FAILED tests/integration/test_api_gateway.py::TestApiGateway::test_api_gateway_404 - json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
FAILED tests/integration/test_api_gateway.py::TestApiGateway::test_dead_letter_queue - json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
ERROR tests/integration/test_api_gateway.py::TestApiGateway::test_dead_letter_queue - json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
=========================================================================================== 3 failed, 1 error in 63.11s (0:01:03) ===========================================================================================
(base)  wtromano+training-Admin-OneClick [wtromano@mbp apigw-sqs-lambda-sqs]$ 
(base)  wtromano+training-Admin-OneClick [wtromano@mbp apigw-sqs-lambda-sqs]$ python -m pytest -s tests/integration -v 
==================================================================================================== test session starts ====================================================================================================
platform darwin -- Python 3.10.6, pytest-7.2.0, pluggy-1.0.0 -- /usr/local/Caskroom/miniforge/base/bin/python
cachedir: .pytest_cache
rootdir: /Users/wtromano/Documents/repo/github/serverless-test-samples/python-test-samples/apigw-sqs-lambda-sqs, configfile: pytest.ini
plugins: socket-0.5.1
collected 3 items                                                                                                                                                                                                           

tests/integration/test_api_gateway.py::TestApiGateway::test_api_gateway_200 FAILED
tests/integration/test_api_gateway.py::TestApiGateway::test_api_gateway_404 FAILED
tests/integration/test_api_gateway.py::TestApiGateway::test_dead_letter_queue FAILED
tests/integration/test_api_gateway.py::TestApiGateway::test_dead_letter_queue ERROR

========================================================================================================== ERRORS ===========================================================================================================
________________________________________________________________________________ ERROR at teardown of TestApiGateway.test_dead_letter_queue _________________________________________________________________________________

self = <class 'test_api_gateway.TestApiGateway'>

    def teardown_class(self) -> None:
        """
        # For tear-down, remove any data injected for the tests
        # purge each queue in case they have messages inside them
        # since purge can be done once every 60 sec, we start with 60 sec sleep
        """
        logging.info("Teardown Phase...")
        time.sleep(60)
    
        # cleaning input queue
>       client = boto3.client("sqs")

tests/integration/test_api_gateway.py:118: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/__init__.py:92: in client
    return _get_default_session().client(*args, **kwargs)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/session.py:299: in client
    return self._session.create_client(
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:951: in create_client
    credentials = self.get_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:509: in get_credentials
    ).load_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:2039: in load_credentials
    creds = provider.load()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:977: in load
    creds_dict = self._retrieve_credentials_using(credential_process)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:1004: in _retrieve_credentials_using
    parsed = botocore.compat.json.loads(stdout.decode('utf-8'))
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/__init__.py:346: in loads
    return _default_decoder.decode(s)
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:337: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x10a81b220>, s = '', idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:355: JSONDecodeError
========================================================================================================= FAILURES ==========================================================================================================
____________________________________________________________________________________________ TestApiGateway.test_api_gateway_200 ____________________________________________________________________________________________

self = <class 'test_api_gateway.TestApiGateway'>

    @classmethod
    def setUp(self) -> None:
        """
        Based on the provided env variable AWS_SAM_STACK_NAME,
        We use the cloudformation API to retrieve the APIGW URL
        """
        stack_name = TestApiGateway.get_stack_name()
    
>       cf_client = boto3.client("cloudformation")

tests/integration/test_api_gateway.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/__init__.py:92: in client
    return _get_default_session().client(*args, **kwargs)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/session.py:299: in client
    return self._session.create_client(
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:951: in create_client
    credentials = self.get_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:509: in get_credentials
    ).load_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:2039: in load_credentials
    creds = provider.load()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:977: in load
    creds_dict = self._retrieve_credentials_using(credential_process)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:1004: in _retrieve_credentials_using
    parsed = botocore.compat.json.loads(stdout.decode('utf-8'))
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/__init__.py:346: in loads
    return _default_decoder.decode(s)
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:337: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x10a81b220>, s = '', idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:355: JSONDecodeError
____________________________________________________________________________________________ TestApiGateway.test_api_gateway_404 ____________________________________________________________________________________________

self = <class 'test_api_gateway.TestApiGateway'>

    @classmethod
    def setUp(self) -> None:
        """
        Based on the provided env variable AWS_SAM_STACK_NAME,
        We use the cloudformation API to retrieve the APIGW URL
        """
        stack_name = TestApiGateway.get_stack_name()
    
>       cf_client = boto3.client("cloudformation")

tests/integration/test_api_gateway.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/__init__.py:92: in client
    return _get_default_session().client(*args, **kwargs)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/session.py:299: in client
    return self._session.create_client(
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:951: in create_client
    credentials = self.get_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:509: in get_credentials
    ).load_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:2039: in load_credentials
    creds = provider.load()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:977: in load
    creds_dict = self._retrieve_credentials_using(credential_process)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:1004: in _retrieve_credentials_using
    parsed = botocore.compat.json.loads(stdout.decode('utf-8'))
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/__init__.py:346: in loads
    return _default_decoder.decode(s)
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:337: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x10a81b220>, s = '', idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:355: JSONDecodeError
___________________________________________________________________________________________ TestApiGateway.test_dead_letter_queue ___________________________________________________________________________________________

self = <class 'test_api_gateway.TestApiGateway'>

    @classmethod
    def setUp(self) -> None:
        """
        Based on the provided env variable AWS_SAM_STACK_NAME,
        We use the cloudformation API to retrieve the APIGW URL
        """
        stack_name = TestApiGateway.get_stack_name()
    
>       cf_client = boto3.client("cloudformation")

tests/integration/test_api_gateway.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/__init__.py:92: in client
    return _get_default_session().client(*args, **kwargs)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/boto3/session.py:299: in client
    return self._session.create_client(
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:951: in create_client
    credentials = self.get_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/session.py:509: in get_credentials
    ).load_credentials()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:2039: in load_credentials
    creds = provider.load()
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:977: in load
    creds_dict = self._retrieve_credentials_using(credential_process)
/usr/local/Caskroom/miniforge/base/lib/python3.10/site-packages/botocore/credentials.py:1004: in _retrieve_credentials_using
    parsed = botocore.compat.json.loads(stdout.decode('utf-8'))
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/__init__.py:346: in loads
    return _default_decoder.decode(s)
/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:337: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x10a81b220>, s = '', idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

/usr/local/Caskroom/miniforge/base/lib/python3.10/json/decoder.py:355: JSONDecodeError
================================================================================================== short test summary info ==================================================================================================
FAILED tests/integration/test_api_gateway.py::TestApiGateway::test_api_gateway_200 - json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
FAILED tests/integration/test_api_gateway.py::TestApiGateway::test_api_gateway_404 - json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
FAILED tests/integration/test_api_gateway.py::TestApiGateway::test_dead_letter_queue - json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
ERROR tests/integration/test_api_gateway.py::TestApiGateway::test_dead_letter_queue - json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
=========================================================================================== 3 failed, 1 error in 62.63s (0:01:02) ===========================================================================================
(base)  wtromano+training-Admin-OneClick [wtromano@mbp apigw-sqs-lambda-sqs]$ ```

Reply:
Verified with TOM, solved by using VENV